drivers/serial/uart_smartbond.c

```diff
@@ -0,0 +1,506 @@
+#include <zephyr/spinlock.h>
+#include <DA1469xAB.h>
+
+#define DT_DRV_COMPAT renesas_smartbond_uart
+
+struct uart_smartbond_baudrate_cfg {
+	uint32_t baudrate;
+	/* DLH=cfg[23:16] DLL=cfg[15:8] DLF=cfg[7:0] */
+	uint32_t cfg;
+};
+
+static const struct uart_smartbond_baudrate_cfg uart_smartbond_baudrate_table[] = {
+	{ 2000000, 0x00000100 },
+	{ 1000000, 0x00000200 },
+	{  921600, 0x00000203 },
+	{  500000, 0x00000400 },
+	{  230400, 0x0000080b },
+	{  115200, 0x00001106 },
+	{   57600, 0x0000220c },
+	{   38400, 0x00003401 },
+	{   28800, 0x00004507 },
+	{   19200, 0x00006803 },
+	{   14400, 0x00008a0e },
+	{    9600, 0x0000d005 },
+	{    4800, 0x0001a00b },
+};
+
+struct uart_smartbond_cfg {
+	UART2_Type *regs;
+	int periph_id;
+	const struct pinctrl_dev_config *pcfg;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct uart_smartbond_data {
+	struct uart_config current_config;
+	struct k_spinlock lock;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+	uint32_t flags;
+	uint8_t rx_enabled;
+	uint8_t tx_enabled;
+#endif
+};
+
+static int uart_smartbond_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if ((config->regs->UART2_USR_REG & UART2_UART2_USR_REG_UART_RFNE_Msk) == 0) {
+		k_spin_unlock(&data->lock, key);
+		return -1;
+	}
+
+	*p_char = config->regs->UART2_RBR_THR_DLL_REG;
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static void uart_smartbond_poll_out(const struct device *dev, unsigned char out_char)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	while (!(config->regs->UART2_USR_REG & UART2_UART2_USR_REG_UART_TFNF_Msk)) {
+		/* Wait until FIFO has free space */
+	}
+
+	config->regs->UART2_RBR_THR_DLL_REG = out_char;
+
+	k_spin_unlock(&data->lock, key);
+}
+
+static int uart_smartbond_configure(const struct device *dev,
+				  const struct uart_config *cfg)
+{
+	const struct uart_smartbond_cfg *config = dev->config;
+	struct uart_smartbond_data *data = dev->data;
+	uint32_t baudrate_cfg = 0;
+	k_spinlock_key_t key;
+	uint32_t reg_val;
+	int err;
+	int i;
+
+	if ((cfg->parity != UART_CFG_PARITY_NONE && cfg->parity != UART_CFG_PARITY_ODD &&
+	     cfg->parity != UART_CFG_PARITY_EVEN) ||
+	    (cfg->stop_bits != UART_CFG_STOP_BITS_1 && cfg->stop_bits != UART_CFG_STOP_BITS_2) ||
+	    (cfg->data_bits != UART_CFG_DATA_BITS_5 && cfg->data_bits != UART_CFG_DATA_BITS_6 &&
+	     cfg->data_bits != UART_CFG_DATA_BITS_7 && cfg->data_bits != UART_CFG_DATA_BITS_8) ||
+	    (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE &&
+	     cfg->flow_ctrl != UART_CFG_FLOW_CTRL_RTS_CTS)) {
+		return -ENOTSUP;
+	}
+
+	/* Flow control is not supported on UART */
```

Comment:
not supported on `uart0` I guess? if that's the case, it would be better to specify this in DT, e.g. using a boolean option `flow-control-supported;` in nodes that support it. See for example https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi#L367 or https://github.com/zephyrproject-rtos/zephyr/pull/46984/commits/2e7304abea25d134ca5ba937f841f685c52b7fb8